core/server/worker/src/test/java/alluxio/worker/block/RegisterStreamerTest.java

```diff
@@ -0,0 +1,402 @@
+          new ImmutableMap.Builder<PropertyKey, Object>()
+          // set response timeout short to ensure that tests don't take too long
+          .put(PropertyKey.WORKER_REGISTER_STREAM_RESPONSE_TIMEOUT, 500)
+          .put(PropertyKey.WORKER_REGISTER_STREAM_DEADLINE, 2000)
+          .put(PropertyKey.WORKER_REGISTER_STREAM_COMPLETE_TIMEOUT, 500)
+          // set batch size to 1 so that the stream will send a request
+          // for every block, used for easily control the concurrency of the stream
+          .put(PropertyKey.WORKER_REGISTER_STREAM_BATCH_SIZE, 1)
+          // we don't handle authentication in testing
+          .put(PropertyKey.SECURITY_AUTHENTICATION_TYPE, AuthType.NOSASL)
+          .build(),
+          Configuration.modifiableGlobal());
+
+  private GrpcServer mServer;
+  private GrpcChannel mChannel;
+
+  @Test(timeout = 5000)
+  public void testRegisterTimeOutConcurrentRequest() throws Exception {
+    // create a server that does not respond to requests
+    createMockService(
+        new BlockMasterWorkerServiceGrpc.BlockMasterWorkerServiceImplBase() {
+          @Override
+          public StreamObserver<RegisterWorkerPRequest> registerWorkerStream(
+              StreamObserver<RegisterWorkerPResponse> responseObserver) {
+            return new StreamObserver<RegisterWorkerPRequest>() {
+              @Override
+              public void onNext(RegisterWorkerPRequest value) {
+              }
+
+              @Override
+              public void onError(Throwable t) {
+              }
+
+              @Override
+              public void onCompleted() {
+              }
+            };
+          }
+        });
+
+    // here we send 5 concurrent requests on the fly
+    // the streamer should throw DeadlineExceededException
+    // because it hasn't received enough response before sending further request
+    RegisterStreamer streamer = createMockRegisterStreamer(mChannel, 1L, 5);
+    assertThrows(DeadlineExceededException.class, streamer::registerWithMaster);
+  }
+
+  @Test(timeout = 5000)
+  public void testRegisterTimeOutSingleRequest() throws Exception {
+    createMockService(
+        new BlockMasterWorkerServiceGrpc.BlockMasterWorkerServiceImplBase() {
+          @Override
+          public StreamObserver<RegisterWorkerPRequest> registerWorkerStream(
+              StreamObserver<RegisterWorkerPResponse> responseObserver) {
+            return new StreamObserver<RegisterWorkerPRequest>() {
+              @Override
+              public void onNext(RegisterWorkerPRequest value) {
+              }
+
+              @Override
+              public void onError(Throwable t) {
+              }
+
+              @Override
+              public void onCompleted() {
+              }
+            };
+          }
+        });
+
+    // the streamer should throw DeadlineExceededException
+    // because the request gets no response
+    RegisterStreamer streamer = createMockRegisterStreamer(mChannel, 0L, 1);
```

Comment:
Why workerId == 0?

Reply:
It shouldn't be. The parameter exists so the test methods can verify the streamer sends the correct workerId. It should be nonzero so as not to be confused with empty protobuf value.